File: backend/schemas/scrape.py
from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel


class ScrapeRunRead(BaseModel):
    id: int
    scraper_type: str
    trigger: str
    schedule_id: Optional[int] = None
    started_at: datetime
    finished_at: Optional[datetime] = None
    status: str
    embedding_status: str
    profiles_requested: int
    items_fetched: int
    error_message: Optional[str] = None
    embedding_error_message: Optional[str] = None
    resume_detected: bool = False
    apify_posts_actor_id: Optional[str] = None
    apify_posts_run_id: Optional[str] = None
    apify_posts_dataset_id: Optional[str] = None
    apify_posts_started_at: Optional[datetime] = None
    apify_posts_finished_at: Optional[datetime] = None
    apify_posts_status: Optional[str] = None
    apify_profiles_actor_id: Optional[str] = None
    apify_profiles_run_id: Optional[str] = None
    apify_profiles_dataset_id: Optional[str] = None
    apify_profiles_started_at: Optional[datetime] = None
    apify_profiles_finished_at: Optional[datetime] = None
    apify_profiles_status: Optional[str] = None
    apify_stage_history: Optional[str] = None

    model_config = {"from_attributes": True}


class ValidateHandlesRequest(BaseModel):
    handles: list[str]
    apify_token: Optional[str] = None


class HandleValidationItemRead(BaseModel):
    submitted_handle: str
    normalized_handle: str
    status: str              # 'FOUND' | 'NOT_FOUND' | 'ERROR'
    instagram_id: Optional[str] = None
    current_handle: Optional[str] = None
    source: Optional[str] = None       # 'database' | 'apify_lookup'
    instagram_url: str
    error_message: Optional[str] = None


class ValidateHandlesResponseRead(BaseModel):
    total: int
    found_count: int
    not_found_count: int
    error_count: int
    results: list[HandleValidationItemRead]


class ScrapeRequest(BaseModel):
    scraper_type: str                       # 'posts' | 'profiles'
    usernames: Optional[list[str]] = None   # override profiles file
    batch_mode: bool = False
    results_limit: int = 100
    only_posts_newer_than: Optional[str] = None
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    data_detail_level: Literal["basicData", "detailedData"] = "basicData"
    enable_embeddings: bool = True
    apify_token: Optional[str] = None


class CombinedScrapeRequest(BaseModel):
    """Request for combined profile + post scrape"""
    usernames: Optional[list[str]] = None   # override profiles file
    batch_mode: bool = False
    results_limit: int = 100
    only_posts_newer_than: Optional[str] = None
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    data_detail_level: Literal["basicData", "detailedData"] = "basicData"
    enable_embeddings: bool = True
    apify_token: Optional[str] = None


class ScrapeStartRead(BaseModel):
    status: str
    profiles_count: int
    run_id: int
    action: Optional[str] = None


class ApifyRefetchRequest(BaseModel):
    stage: Literal["posts", "profiles"]
    include_logs: bool = True


class ApifyRefetchRead(BaseModel):
    run_id: int
    stage: Literal["posts", "profiles"]
    apify_run_id: str
    apify_dataset_id: str
    apify_status: Optional[str] = None
    items_count: int = 0
    logs_count: int = 0
    status: str = "refetched"


class ScrapeRunListResponse(BaseModel):
    items: list[ScrapeRunRead]
    total: int


class ProfilesSourceRead(BaseModel):
    usernames: list[str]


class ProfilesSourceUpdate(BaseModel):
    usernames: list[str]


class ScrapeProfileRead(BaseModel):
    id: int
    username: str
    category: Optional[str] = None
    grade: Optional[str] = None
    position: int
    profile_pic_url: Optional[str] = None

    model_config = {"from_attributes": True}


class ScrapeProfileCreate(BaseModel):
    username: str
    category: Optional[str] = None
    grade: Optional[str] = None


class ScrapeProfileBulkCreate(BaseModel):
    profiles: list[ScrapeProfileCreate]


class ScrapeProfileBulkResult(BaseModel):
    created: list[ScrapeProfileRead]
    skipped_existing: list[str] = []


class ScrapeProfileUpdate(BaseModel):
    username: Optional[str] = None
    category: Optional[str] = None
    grade: Optional[str] = None


class ScrapeDbUpdateStatus(BaseModel):
    posts_rows: int = 0
    profile_snapshots_rows: int = 0
    profiles_touched: int = 0
    missing_usernames: list[str] = []


class ScrapeProfileFailureRead(BaseModel):
    username: str
    attempt_count: int = 0
    error_message: Optional[str] = None
    failure_category: str = "unknown"
    retryable: bool = False
    retries_left: int = 0


class ScrapeProfileAttemptRead(BaseModel):
    username: str
    status: str
    attempt_count: int = 0


class ScrapeProfileProgressRead(BaseModel):
    total_profiles: int = 0
    completed_count: int = 0
    pending_count: int = 0
    failed_count: int = 0
    retryable_failed_count: int = 0
    terminal_failed_count: int = 0
    running_count: int = 0
    completed_profiles: list[str] = []
    pending_profiles: list[str] = []
    failed_profiles: list[ScrapeProfileFailureRead] = []
    zero_posts_profiles: list[str] = []
    profile_attempts: list[ScrapeProfileAttemptRead] = []
    server_failure_message: Optional[str] = None


class ScrapeProfileProgressRowRead(BaseModel):
    username: str
    status: str
    attempt_count: int = 0
    items_fetched: int = 0
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    last_checkpoint_at: Optional[datetime] = None


class ScrapeProfileProgressListResponse(BaseModel):
    items: list[ScrapeProfileProgressRowRead]
    total: int
    limit: int
    offset: int


class ScrapeStatusRead(BaseModel):
    run: Optional[ScrapeRunRead] = None
    progress_pct: int = 0
    db_updates: ScrapeDbUpdateStatus = ScrapeDbUpdateStatus()
    profile_progress: ScrapeProfileProgressRead = ScrapeProfileProgressRead()
    resume_detected: bool = False
    logs: list[str] = []


class CompareSummaryRead(BaseModel):
    run_a_profile_snapshot_rows: int = 0
    run_b_profile_snapshot_rows: int = 0
    run_a_latest_posts_rows: int = 0
    run_b_latest_posts_rows: int = 0
    common_profiles: int = 0
    new_profiles: int = 0
    missing_profiles: int = 0
    net_followers_delta: int = 0
    common_latest_posts: int = 0
    new_latest_posts: int = 0
    missing_latest_posts: int = 0
    net_likes_delta: int = 0


class ProfileDeltaRead(BaseModel):
    profile_id: str
    followers_run_a: int | None = None
    followers_run_b: int | None = None
    follows_run_a: int | None = None
    follows_run_b: int | None = None
    posts_run_a: int | None = None
    posts_run_b: int | None = None
    followers_delta: int = 0
    follows_delta: int = 0
    posts_delta: int = 0
    change_type: Literal["common", "new", "missing"]


class LatestPostDeltaRead(BaseModel):
    profile_id: str
    owner_username: str | None = None
    url: str
    likes_run_a: int | None = None
    likes_run_b: int | None = None
    comments_run_a: int | None = None
    comments_run_b: int | None = None
    views_run_a: int | None = None
    views_run_b: int | None = None
    likes_delta: int = 0
    comments_delta: int = 0
    views_delta: int = 0
    change_type: Literal["common", "new", "missing"]


class InsightRead(BaseModel):
    title: str
    value: str
    detail: str
    tone: Literal["positive", "negative", "neutral"] = "neutral"


class RunComparisonRead(BaseModel):
    run_a: ScrapeRunRead
    run_b: ScrapeRunRead
    summary: CompareSummaryRead
    profile_deltas: list[ProfileDeltaRead]
    latest_post_deltas: list[LatestPostDeltaRead]
    insights: list[InsightRead]

File: frontend/src/api/scrape.ts
import { API_URL } from "../config"
import type { RunComparison, ScrapeRun } from "../types/schedule"

export interface ScrapeRequest {
  scraper_type: string
  usernames?: string[]
  batch_mode?: boolean
  results_limit?: number
  only_posts_newer_than?: string
  date_from?: string
  date_to?: string
  data_detail_level?: "basicData" | "detailedData"
  enable_embeddings?: boolean
  apify_token?: string
}

export interface CombinedScrapeRequest {
  usernames?: string[]
  batch_mode?: boolean
  results_limit?: number
  only_posts_newer_than?: string
  date_from?: string
  date_to?: string
  data_detail_level?: "basicData" | "detailedData"
  enable_embeddings?: boolean
  apify_token?: string
}

export interface ScrapeStartResponse {
  status: string
  profiles_count: number
  run_id: number
  action?: string
}

export interface HandleValidationItem {
  submitted_handle: string
  normalized_handle: string
  status: "FOUND" | "NOT_FOUND" | "ERROR"
  instagram_id?: string | null
  current_handle?: string | null
  source?: "database" | "apify_lookup" | null
  instagram_url: string
  error_message?: string | null
}

export interface ValidateHandlesResponse {
  total: number
  found_count: number
  not_found_count: number
  error_count: number
  results: HandleValidationItem[]
}

export interface ProfilesSourceResponse {
  usernames: string[]
}

export interface ScrapeDbUpdateStatus {
  posts_rows: number
  profile_snapshots_rows: number
  profiles_touched: number
  missing_usernames: string[]
}

export interface ScrapeProfileFailure {
  username: string
  attempt_count: number
  error_message?: string
  failure_category: string
  retryable: boolean
  retries_left: number
}

export interface ScrapeProfileAttempt {
  username: string
  status: string
  attempt_count: number
}

export interface ScrapeProfileProgressRow {
  username: string
  status: string
  attempt_count: number
  items_fetched: number
  error_message?: string
  started_at?: string
  finished_at?: string
  last_checkpoint_at?: string
}

export interface ScrapeProfileProgressListResponse {
  items: ScrapeProfileProgressRow[]
  total: number
  limit: number
  offset: number
}

export interface ScrapeProfileProgress {
  total_profiles: number
  completed_count: number
  pending_count: number
  failed_count: number
  retryable_failed_count: number
  terminal_failed_count: number
  running_count: number
  completed_profiles: string[]
  pending_profiles: string[]
  failed_profiles: ScrapeProfileFailure[]
  zero_posts_profiles: string[]
  profile_attempts: ScrapeProfileAttempt[]
  server_failure_message?: string
}

export interface ScrapeStatusResponse {
  run: ScrapeRun | null
  progress_pct: number
  db_updates: ScrapeDbUpdateStatus
  profile_progress?: ScrapeProfileProgress
  resume_detected?: boolean
  logs: string[]
}

export interface ApifyRefetchResponse {
  run_id: number
  stage: "posts" | "profiles"
  apify_run_id: string
  apify_dataset_id: string
  apify_status?: string
  items_count: number
  logs_count: number
  status: string
}

export const triggerScrape = (body: ScrapeRequest): Promise<ScrapeStartResponse> =>
  fetch(`${API_URL}/api/scrape/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then(r => r.json())

export const validateHandles = (handles: string[], apifyToken?: string): Promise<ValidateHandlesResponse> =>
  fetch(`${API_URL}/api/scrape/validate-handles`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ handles, apify_token: apifyToken }),
  }).then(r => r.json())

export const triggerCombinedScrape = (body: CombinedScrapeRequest): Promise<ScrapeStartResponse> =>
  fetch(`${API_URL}/api/scrape/run/combined`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then(r => r.json())

export const fetchRuns = (params: Record<string, string | number | undefined> = {}): Promise<{ items: ScrapeRun[]; total: number }> => {
  const qs = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) qs.set(k, String(v))
  }
  return fetch(`${API_URL}/api/scrape/runs?${qs}`).then(r => r.json())
}

export const fetchProfilesSource = (): Promise<ProfilesSourceResponse> =>
  fetch(`${API_URL}/api/scrape/profiles-source`).then(r => r.json())

export const updateProfilesSource = (usernames: string[]): Promise<ProfilesSourceResponse> =>
  fetch(`${API_URL}/api/scrape/profiles-source`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ usernames }),
  }).then(r => r.json())

export const fetchScrapeStatus = (runId?: number): Promise<ScrapeStatusResponse> => {
  const qs = new URLSearchParams()
  if (runId !== undefined) qs.set("run_id", String(runId))
  return fetch(`${API_URL}/api/scrape/status?${qs}`).then(r => r.json())
}

export const fetchRunProfileProgress = (
  runId: number,
  params: Record<string, string | number | undefined> = {},
): Promise<ScrapeProfileProgressListResponse> => {
  const qs = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== "") qs.set(k, String(v))
  }
  return fetch(`${API_URL}/api/scrape/runs/${runId}/profile-progress?${qs.toString()}`).then(r => r.json())
}

export const skipEmbedding = (runId: number): Promise<ScrapeRun> =>
  fetch(`${API_URL}/api/scrape/runs/${runId}/skip-embedding`, { method: "PATCH" }).then(r => {
    if (!r.ok) return r.json().then(e => Promise.reject(new Error(e.detail ?? "Failed")))
    return r.json()
  })

export const resumePendingPosts = (runId: number): Promise<ScrapeStartResponse> =>
  fetch(`${API_URL}/api/scrape/runs/${runId}/resume-pending-posts`, {
    method: "POST",
  }).then(async r => {
    if (!r.ok) {
      const err = await r.json().catch(() => ({ detail: "Failed to resume pending profiles" }))
      throw new Error(err.detail ?? "Failed to resume pending profiles")
    }
    return r.json()
  })

export const refetchRunFromApify = (
  runId: number,
  stage: "posts" | "profiles",
  includeLogs = true,
): Promise<ApifyRefetchResponse> =>
  fetch(`${API_URL}/api/scrape/runs/${runId}/apify-refetch`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ stage, include_logs: includeLogs }),
  }).then(async r => {
    if (!r.ok) {
      const err = await r.json().catch(() => ({ detail: "Failed to refetch from Apify" }))
      throw new Error(err.detail ?? "Failed to refetch from Apify")
    }
    return r.json()
  })

export const fetchRunComparison = (
  runAId: number,
  runBId: number,
  profileLimit = 50,
  latestPostLimit = 50,
): Promise<RunComparison> => {
  const qs = new URLSearchParams({
    run_a_id: String(runAId),
    run_b_id: String(runBId),
    profile_limit: String(profileLimit),
    latest_post_limit: String(latestPostLimit),
  })
  return fetch(`${API_URL}/api/scrape/runs/compare?${qs}`).then(async r => {
    if (!r.ok) {
      const err = await r.json().catch(() => ({ detail: "Failed to compare runs" }))
      throw new Error(err.detail ?? "Failed to compare runs")
    }
    return r.json()
  })
}
